### DISCUSSION TITLE
Silhouette highlighting

### DISCUSSION THREAD

**Discussion by trevorsandy** [leozide/leocad#700] [Ideas]
Hi Leo. 

What are your thoughts on enabling silhouette highlighting as shown in step 2 below ? This format is more elegant than the current shotgun approach. 

![lego highlight](https://user-images.githubusercontent.com/13388970/115157777-5d3b5580-a08b-11eb-9774-e94e3d6c20a7.jpg)

If you care to highlight the best approach for LeoCAD, I will take a look.

Cheers,

**Comment by rsbx** [comment]
+1

**Comment by leozide** [comment]
I think it's a good idea, there was a bug about it I think.

I've never done much research on this but you probably need to render the highlighted parts twice, first draw the silhouette by expanding the verts in the direction of the normals (scaled by a camera factor so the results are constant in screen space), and probably only draw the back faces. Then draw the regular part on top.

Maybe stay away from the post processing algorithms for portability.

**Reply by rsbx** [reply]
#461

**Comment by trevorsandy** [comment]
@rsbx your write up looks good. The second scenario seems closer to what Leo is suggesting.

Cheers,

**Comment by rsbx** [comment]
@trevorsandy the second method ("fringe") is the best I've come up with that's consistent with the silhouetting artifacts I've found in LEGO instructions. It has the advantage that much of the work happens on the GPU. The disadvantages I've identified are
- limited silhouette thickness
- silhouetting happens in _model space_ so image silhouette thickness decreases with camera distance
- silhouetting artifacts are unavoidable.

The first method ("glow") is what you'd get if the highlighted parts glowed or by meticulously hand drawing the silhouette. It's advantages and disadvantages are, pretty much, the opposite of the second method.

Upon reading what I wrote a year ago, I think the description of the needed masking is not quite correct. I'll try to post an updated, more OGL'y, description later today.
